File: solutions/2026-06-22/python/2026-06-22-maximum-number-of-balloons.py
```python
# Problem: Maximum Number of Balloons
# Link: https://leetcode.com/problems/maximum-number-of-balloons/
# Approach:
# The problem asks us to find the maximum number of times we can form the word "balloon"
# using the characters from a given text. We can only use each character once.
# To solve this, we need to count the frequency of each character required for "balloon"
# ('b', 'a', 'l', 'o', 'n') in the input text.
# For each character in "balloon", we determine how many times it can be formed
# based on its available count in the text.
# Specifically, 'b', 'a', and 'n' are needed once per "balloon".
# 'l' and 'o' are needed twice per "balloon".
#
# The limiting factor for the number of "balloon" instances will be the character
# that is least available relative to its requirement.
#
# We can use a hash map (dictionary in Python) to store the counts of characters in the input text.
# Then, we iterate through the required characters of "balloon":
# - For 'b', 'a', 'n': The number of "balloon"s is limited by the count of this character in `text`.
# - For 'l', 'o': Since each "balloon" needs two 'l's and two 'o's, the number of "balloon"s
#   is limited by `count_of_char // 2`.
#
# The minimum of these limitations for all characters will be our answer.
#
# Time Complexity: O(N), where N is the length of the input string `text`.
# We iterate through the input string once to count character frequencies.
# Then, we iterate through a fixed number of characters ('b', 'a', 'l', 'o', 'n')
# to determine the minimum instances.
#
# Space Complexity: O(1). We are using a hash map to store character counts.
# Since the input string consists only of lowercase English letters, the size of the
# hash map will be at most 26, which is a constant.

from collections import Counter

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        # Count the frequency of each character in the input text.
        # The Counter object provides a convenient way to do this.
        char_counts = Counter(text)

        # The word "balloon" requires the following characters with these frequencies:
        # 'b': 1
        # 'a': 1
        # 'l': 2
        # 'o': 2
        # 'n': 1

        # Calculate the maximum number of balloons based on the availability of each character.
        # For characters 'b', 'a', 'n', we need one of each per balloon.
        # So, the number of balloons is limited by their count directly.
        max_b = char_counts.get('b', 0)
        max_a = char_counts.get('a', 0)
        max_n = char_counts.get('n', 0)

        # For characters 'l' and 'o', we need two of each per balloon.
        # So, the number of balloons is limited by half of their count.
        max_l = char_counts.get('l', 0) // 2
        max_o = char_counts.get('o', 0) // 2

        # The overall maximum number of balloons we can form is limited by the minimum
        # of the calculated maximums for each required character.
        # We use `min()` to find the bottleneck.
        return min(max_b, max_a, max_l, max_o, max_n)

```

File: solutions/2026-06-22/java/2026-06-22-maximum-number-of-balloons.java
// Problem: Maximum Number of Balloons
// Problem Summary: Determine how many times the word "balloon" can be formed using characters from a given text.
// Link: https://leetcode.com/problems/maximum-number-of-balloons/
//
// Approach:
// To solve this problem, we need to count the frequency of each character in the input `text` and compare it with the required frequency of characters in the word "balloon".
// The word "balloon" requires:
// - 1 'b'
// - 1 'a'
// - 2 'l's
// - 2 'o's
// - 1 'n'
//
// We can use a frequency map (or an array of size 26 for lowercase English letters) to store the counts of characters in `text`.
// After counting, we calculate how many "balloons" can be formed based on the available characters.
// For 'b', 'a', and 'n', the number of balloons is limited by their count in `text`.
// For 'l' and 'o', since we need two of each per "balloon", the number of balloons is limited by half of their count in `text` (integer division).
// The minimum of these limitations will be our final answer.
//
// Time Complexity: O(N), where N is the length of the input `text`.
// We iterate through the `text` once to build the frequency map. Then, we perform a constant number of operations (checking counts for 5 distinct characters) to determine the maximum number of balloons.
//
// Space Complexity: O(1).
// The space used by the frequency map (or array) is constant, as it only stores counts for 26 lowercase English letters, regardless of the input string's length.

class Solution {
    public int maxNumberOfBalloons(String text) {
        // Frequency map to store the count of each character in the input text.
        // Using an array of size 26 for lowercase English letters ('a' through 'z').
        int[] charCounts = new int[26];

        // Iterate through the input string to populate the frequency map.
        for (char c : text.toCharArray()) {
            // Increment the count for the corresponding character.
            // 'c - 'a'' gives the index for the character (e.g., 'a' -> 0, 'b' -> 1).
            charCounts[c - 'a']++;
        }

        // Define the required counts for each character in the word "balloon".
        // b: 1, a: 1, l: 2, o: 2, n: 1
        int countB = charCounts['b' - 'a']; // Count of 'b'
        int countA = charCounts['a' - 'a']; // Count of 'a'
        int countL = charCounts['l' - 'a']; // Count of 'l'
        int countO = charCounts['o' - 'a']; // Count of 'o'
        int countN = charCounts['n' - 'a']; // Count of 'n'

        // Calculate the maximum number of balloons that can be formed.
        // For 'b', 'a', and 'n', the number of balloons is directly limited by their counts.
        // For 'l' and 'o', since we need two of each per balloon, the number of balloons is limited by half of their counts (integer division).
        // The overall maximum number of balloons is the minimum of these individual limits.
        int maxBalloons = Math.min(countB, countA); // Limited by 'b' and 'a'
        maxBalloons = Math.min(maxBalloons, countN); // Limited by 'b', 'a', and 'n'
        maxBalloons = Math.min(maxBalloons, countL / 2); // Limited by 'b', 'a', 'n', and 'l' (need 2 'l's per balloon)
        maxBalloons = Math.min(maxBalloons, countO / 2); // Limited by all characters (need 2 'o's per balloon)

        // Return the calculated maximum number of balloons.
        return maxBalloons;
    }
}
